---
title: "Lab 10 : National Ecological Observatory Network (NEON) Projects"
output:
  html_document:
    toc: true
    toc_depth: 4
    theme: united
    highlight: tango
bibliography: references.bib
---

## Learning objectives

* Revisiting the National Ecological Observatory Network 
* Revisiting the JGI and the Integrated Microbial Genomes and Microbiomes (IMG/MER)
* Adding references to your report
* Starting your NEON project

## Background

We will revisit following in class

* [NEON](https://www.neonscience.org/)
* [NEON Field Sites](https://www.neonscience.org/field-sites)
* [Integrated Microbial Genomes and Microbiomes(IMG/MER)](https://img.jgi.doe.gov/cgi-bin/mer/main.cgi)


## Adding references to your report

You can automatically generate citations and a bibliography in a number of styles. In order to use this feature, you will need to specify a bibliography file using the bibliography metadata field in a YAML metadata section. For example:

```
---
title: "Sample Document"
output: html_document
bibliography: bibliography.bib
---
```
The bibliography may have any of these formats:

Format    |File extension
----------|--------------
MODS |.mods
BibLaTeX|.bib
BibTeX|.bibtex
RIS|.ris
EndNote|.enl
EndNote XML|.xml
ISI|.wos
|MEDLINE|.medline
|Copac|.copac
|JSON citeproc|.json

You can export references from your favorite reference manager. I use the open source [Zotero](www.zotero.org).

### Minimum working example

If you create these two files `paper.Rmd` and `references.bib` you will have a working example to build from
paper.Rmd

```
---
title: 'My Title'
author: "Me me me me!"
date: Aug 31, 2020
output: html_document
bibliography: references.bib
---

Application written in R [@RCoreTeam] using the Shiny framework [@Chang2015].

# REFERENCES
```

references.bib (Make this file as a text file (File -> New File -> text file))

```
@Misc{Chang2015,
  Title                    = {shiny: Web Application Framework for R. R package version 0.12.1},

  Author                   = {Chang, W. and Cheng, J. and Allaire, JJ. and Xie, Y. and McPherson, J. },
  Year                     = {2015},

  Type                     = {Computer Program},
  Url                      = {http://CRAN.R-project.org/package=shiny}
}


@Article{RCoreTeam,
  Title                    = {R: A Language and Environment for Statistical Computing},
  Author                   = {{R Core Team}},
  Year                     = {2015},

  Type                     = {Journal Article},
  Url                      = {http://www.R-project.org}
}
```

Here is a short example of a manuscript style report [Reproducible Papers with RMarkdown](https://nceas.github.io/oss-lessons/reproducible-papers-with-rmd/reproducible-papers-with-rmd.html)


## Integrated Microbial Genomes and Microbiomes(IMG/MER)

* To create an account go to [Integrated Microbial Genomes and Microbiomes(IMG/MER)](https://img.jgi.doe.gov/cgi-bin/mer/main.cgi). Register for an account using your UMass email. For funding source use DOE Biological and Environmental Research (BER)

The GOLD Study ID for the project is [Gs0161344](https://gold.jgi.doe.gov/study?id=Gs0161344)

We will go through working with the data associated with the metagenomes today.

### Project Groups

```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(knitr)
library(tidyverse)
kable(
project_groups <- read_csv("data/NEON/project_groups.csv") 
)
```

## Lab 10 Report

You can work together with your group, but each of you should submit a lab report to Canvas. For today's report include:

### Exercise 1

* A description of your NEON site including a link to the NEON site information as reference and photos.

### Exercise 2

* An overview of your Phylum/Class including references.

### Exercise 3

Using the table from Labs 8 and 9 analyze

* The MAGs at your site including the taxomic breakdown. Use both tables and graphs.
* The sites your Phylum/Class is found at. Include tables and graphs with lower level taxonomic information.

### Exercise 4

From IMG create table from IMG with the metagenome information including Ecosystem subtype, latitude, longitude, Genome Size, Gene Count. Just display the table in your Lab 10 report.  We will work more with this table next week. 

To do this

* Go to (IMG/MER)](https://img.jgi.doe.gov/cgi-bin/mer/main.cgi)
* Search in IMG using the `Quick Genome Search` with the Study ID `Gs0161344` to find metagenomes associated with our project. 
* Click on the `176` Hits.
* Scroll down and check the boxes corresponding to Ecosystem Categories (5), latitude and longitude. The the default checked boxes check
* Click on `Redisplay`
* Scroll up click on `Select All` (There should be 176 metagenomes)
* Click on `Export`
* This will create a csv file for download. Put this in your data/NEON folder
* Read in the file using `read_tsv` because although it says it is a csv file it is actually a tsv (tab separated value) file. 
* Filter the table to include only metagenomes from your site and metagenomes that are `re-annotation`.
* Display the table using kable or DT(datatable) in your report.

## Acknowledgements

Our labs are written in R [@RCoreTeam] and past labs have used th Shiny framework [@Chang2015].

## References
